router/core/timout_transport_test.go

```diff
@@ -0,0 +1,180 @@
+			require.Nil(t, resp) // No response due to timeout
+		})
+	})
+
+	t.Run("ResponseHeaderTimeout exceeded", func(t *testing.T) {
+		t.Parallel()
+
+		transportOpts.SubgraphMap[testSubgraphKey].ResponseHeaderTimeout = 3 * time.Millisecond
+
+		headerTimeoutServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			time.Sleep(5 * time.Millisecond) // Delayed header response
+			w.WriteHeader(http.StatusOK)
+		}))
+		defer headerTimeoutServer.Close()
+
+		rqCtx := &requestContext{
+			subgraphResolver: NewSubgraphResolver([]Subgraph{{Name: "test", UrlString: headerTimeoutServer.URL}}),
+		}
+
+		req := httptest.NewRequest("GET", headerTimeoutServer.URL, nil)
+		req = req.WithContext(withRequestContext(req.Context(), rqCtx))
+
+		timeoutTransport := NewTimeoutTransport(
+			transportOpts,
+			http.DefaultTransport,
+			zap.NewNop(),
+			http.ProxyFromEnvironment,
+		)
+
+		resp, err := timeoutTransport.RoundTrip(req)
+		require.NotNil(t, err)
+		require.ErrorAs(t, err, &context.DeadlineExceeded)
+		require.Nil(t, resp)
+	})
+
+	t.Run("TLSHandshakeTimeout exceeded", func(t *testing.T) {
+		t.Parallel()
+
+		tlsServer := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			w.WriteHeader(http.StatusOK)
+		}))
+		tlsServer.TLS = &tls.Config{}
+		tlsServer.StartTLS()
+		defer tlsServer.Close()
+
+		transportOpts.SubgraphMap[testSubgraphKey].TLSHandshakeTimeout = 2 * time.Millisecond
+
+		rqCtx := &requestContext{
+			subgraphResolver: NewSubgraphResolver([]Subgraph{{Name: "test", UrlString: tlsServer.URL}}),
+		}
+
+		req := httptest.NewRequest("GET", tlsServer.URL, nil)
+		req = req.WithContext(withRequestContext(req.Context(), rqCtx))
+
+		timeoutTransport := NewTimeoutTransport(
+			transportOpts,
+			http.DefaultTransport,
+			zap.NewNop(),
+			http.ProxyFromEnvironment,
+		)
+
+		resp, err := timeoutTransport.RoundTrip(req)
+		require.NotNil(t, err)
+		require.Contains(t, err.Error(), "TLS handshake timeout")
+		require.ErrorAs(t, err, &os.ErrDeadlineExceeded)
+		require.Nil(t, resp)
+	})
+
+	t.Run("DialTimeout exceeded", func(t *testing.T) {
+		t.Parallel()
+
+		transportOpts.SubgraphMap[testSubgraphKey].DialTimeout = 1 * time.Millisecond
+		unreachableServerURL := "http://192.0.2.1" // Reserved IP address unlikely to respond
+
+		rqCtx := &requestContext{
+			subgraphResolver: NewSubgraphResolver([]Subgraph{{Name: testSubgraphKey, UrlString: unreachableServerURL}}),
+		}
+
+		req := httptest.NewRequest("GET", unreachableServerURL, nil)
+		req = req.WithContext(withRequestContext(req.Context(), rqCtx))
+
+		timeoutTransport := NewTimeoutTransport(
+			transportOpts,
+			http.DefaultTransport,
+			zap.NewNop(),
+			http.ProxyFromEnvironment,
+		)
+
+		resp, err := timeoutTransport.RoundTrip(req)
+		require.NotNil(t, err)
+		require.ErrorContains(t, err, "dial tcp")
+		require.ErrorAs(t, err, &os.ErrDeadlineExceeded)
+		require.Nil(t, resp)
+	})
```

Comment:
I found it difficult to make isolated keepalive* and expectedcontinue unit tests.

Reply:
expect_continue_timeout is not relevant for GraphQL. For keepalive you can take inspiration from this snippet

```go
package main

import (
	"io/ioutil"
	"net"
	"net/http"
	"net/http/httptest"
	"sync/atomic"
	"testing"
	"time"
)

func TestIdleConnTimeout(t *testing.T) {
	var connCount int32

	// Create a test server that increments connCount each time a new connection is established.
	server := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte("ok"))
	}))
	
	// Track connections using ConnState
	server.Config.ConnState = func(c net.Conn, s http.ConnState) {
		if s == http.StateNew {
			atomic.AddInt32(&connCount, 1)
		}
	}

	server.StartTLS()
	defer server.Close()

	// Configure a transport with a short idle connection timeout.
	transport := &http.Transport{
		TLSClientConfig: server.Client().Transport.(*http.Transport).TLSClientConfig,
		IdleConnTimeout: 200 * time.Millisecond, // short idle timeout
	}
	client := &http.Client{Transport: transport}

	// Make an initial request to establish a connection.
	resp, err := client.Get(server.URL)
	if err != nil {
		t.Fatalf("failed to make first request: %v", err)
	}
	resp.Body.Close()

	// At this point, one connection should have been established.
	firstConnCount := atomic.LoadInt32(&connCount)
	if firstConnCount != 1 {
		t.Fatalf("expected 1 connection established, got %d", firstConnCount)
	}

	// Wait longer than the idle timeout so the connection goes idle and is closed.
	time.Sleep(500 * time.Millisecond)

	// Make a second request. If the IdleConnTimeout works, a new connection must be established.
	resp2, err := client.Get(server.URL)
	if err != nil {
		t.Fatalf("failed to make second request: %v", err)
	}
	ioutil.ReadAll(resp2.Body)
	resp2.Body.Close()

	// Check the connection count again.
	// We expect another connection to have been created since the first should have been closed due to idle timeout.
	finalConnCount := atomic.LoadInt32(&connCount)
	if finalConnCount <= firstConnCount {
		t.Fatalf("expected a new connection after idle timeout, but connection count did not increase (still %d)", finalConnCount)
	}

	t.Logf("Connection count increased from %d to %d as expected, indicating IdleConnTimeout took effect", firstConnCount, finalConnCount)
}
```